### DISCUSSION TITLE
STM32CubeIDE Integration

### DISCUSSION THREAD

**Discussion by MimitechIndustries** [hathach/tinyusb#633] [Q&A] [answered]
I've been able to build/make the tinyusb examples for various targets (including STM32), but, I am completely lost in figuring out how to integrate it into happily building in an existing project in STM32CubeIDE.  I will probably figure it out eventually and when I do I will post the details into this discussion since there doesn't seem to be a reference anywhere on the general Internet.  If somebody already knows how and wants to assist to help speed things up that would also be cool :) 

Update: I have figured out how to do this finally; however it's 5:30am and I'm going to sleep 🙃 I'll answer the question tomorrow

**Comment by zlittell** [comment]
I have recently done this with microchip IDEs and really meant to make a blog post and just haven't had time. What you should start out doing is either include tinyusb as a sub module and initialize it recursively or copy tinyusb into your project folder. Then I go into an example for the type of device I am making and open the makefile. I then look for all the spots a makefile is included and open those. Then go one by one copying over any include lines that you find throughout these make files into a separate text file. Then in STM32CUBE there should be a way to edit project include folders, start adding in all the directories. Then do the same with all the c files. You want to add then as a link though, I am not sure how stm32cube has you do that. (You just don't want it to copy the file to where you are adding it, which atmel studio will do). 

This is definitely something that is easier to see done than read but I hope this was at least somewhat helpful.

**Reply by MimitechIndustries** [reply]
Thanks for the advice, this method may also be viable.

**Comment by MimitechIndustries** [comment, marked as answer]
The procedure:
* In the graphical chip configurator, enable USB_OTG_HS (or whichever USB peripheral you have under 'Connectivity' section) and set it in "Device_Only" mode.  In the bottom, under NVIC Settings, check the box that is the global USB interrupt.  Make sure you are not using the ST USB middleware -- make sure in the 'Middleware' section under USB_DEVICE and USB_HOST (if you have this section) that all the Class boxes are set to "Disable"
* Check out tinyusb to your project's root -- do not put it in "Middlewares" or "Drivers" since the IDE automatically includes practically *everything* inside these folders and there are all sorts of examples, libraries, etc. that do not target *your* microcontroller and/or board that get included
* Right click the project, go to properties.  in C/C++ General -> Paths and Symbols, in the "Includes" tab add a path, make it a workspace path, point it to the tinyusb/src directory.  Then, in the "Source Location" tab, add folder, and also point it to tinyusb/src
* Add a tusb_config.h to your "Core/Inc" 
```c
// ^^^ <...> omitted stuff above, generic C language boilerplate crap, see any example in tinyusb

#define CFG_TUSB_MCU                 OPT_MCU_STM32H7
#define CFG_TUSB_OS                  OPT_OS_NONE
#define BOARD_DEVICE_RHPORT_SPEED    OPT_MODE_HIGH_SPEED  // 480mbps
#define BOARD_DEVICE_RHPORT_NUM     1
#define CFG_TUSB_RHPORT1_MODE       (OPT_MODE_DEVICE | OPT_MODE_HIGH_SPEED)

// In the case of the STM32H7 with an external HS 480 PHY, you must use root hub port 1 instead of 0
//    0 is for the internal FS 12mbit PHY so you'd use BOARD_DEVICE_RHPORT_NUM set to 0 and CFG_TUSB_RHPORT1_MODE set to (OPT_MODE_DEVICE | OPT_MODE_FULL_SPEED)

// vvv<...> (see tinyusb examples for remainder, configuring USB Class and buffer size basics)
```
* Add a `usb_descriptors.h` and `usb_descriptors.c` file to your Core/Inc and Core/Src folders and populate them with the USB descriptors you need.  Look at the various tinyusb included examples to get started (copy paste one closest to what you need)
* `#include "tusb.h"` in your main.c and call `tusb_init()` after the other init code, then in your application loop, call `tud_task()` 
* in your `<chipname>_it.c` file (in my case `stm32h7xx_it.c`) in Core/Src that was generated by STM32CubeIDE, locate your USB global IRQ handler function and make it instead call `tud_int_handler(BOARD_DEVICE_RHPORT_NUM)` and then `return` on the next line (do this in the 'USER CODE' block so that it doesn't get erased when you update the chip config in the future).  The `HAL_PCD_IRQHandler` function call, if you have one, will not be reachable and that is what we want so that ST's USB handling code isn't called.  

There may be a cleaner way to do this.  It might not be necessary to use the USB initialization code provided by ST at all and let tinyusb handle that; however I haven't tested that yet.  Will update post in the future.

I wrote this from memory, will double check when I make a new project with this procedure sometime soon.

**Reply by hathach** [reply]
thank you for your instruction, that will surely helps lots of other STM32 user :+1: :+1:

**Reply by squintz** [reply]
I tried following Mimitechindustries writeup but I failed and I'm not sure where my problem is. I'm a complete newbie to STM32s and I'm attempting to implement a basic USB HS CDC loopback as practice. I'm working with a Waveshare STM32H743 dev board with the waveshare ULPI usb module. 

I managed to get the CubeMX libraries to enumerate but couldn't open a port without RealTerm complaining. I decided to give TinyUSB a try hoping it would be easier to work with. I'm getting stuck in dcd_synopsys.c on line 471:

`while ((usb_otg->GRSTCTL & USB_OTG_GRSTCTL_AHBIDL) == 0U) {}`

I'm using the same ioc file that came with the waveshare dev board examples which almost works with the STM32 libraries. My code compiles but nothing happens. I'd love to see some screenshots or a video of CubeIDE if someone has a basic working implementation they can share. I have a feeling I'm missing some important stuff from the family.c and board files.

**Reply by hathach** [reply]
how about getting the board to build with current makefile in the repo. It has most of the setup done properly. clock & phy setup in these mcu can get complicated.

PS: please create a new discussion. This one is solved now.

**Reply by WebiusD** [reply]
For people who try to use the STM32 in HOST mode. We need some minor adjustments from the above explanation (which is DEVICE mode). Most of the changes are obvious, while there are some tricky parts as well. I will only list here, whenever we have to deviate from the above description.
1) In the graphical chip configurator select Host_Only (obvious), keep the USB OTG FS global interrupt checked. I powered my USB device externally and needed neither the start of frame bit (SOF) nor vbus sensing pin (VBUS).
2) The tusb_config.h file can mostly stay the same again simply exchange DEVICE with HOST. Since a host needs to receive the descriptors send by the device we need to tell tinyUSB how big that receive buffer shall be.
```
// Set MCU as usual:
#define CFG_TUSB_MCU		OPT_MCU_STM32F4
// Enable Host stack:
#define CFG_TUH_ENABLED       1
// Set root hub port:
#ifndef BOARD_TUH_RHPORT
#define BOARD_TUH_RHPORT      0
// this allows us to call tusb_init() without arguments in main.c, since we already set the mode here:
#define CFG_TUSB_RHPORT0_MODE		(OPT_MODE_HOST | OPT_MODE_FULL_SPEED)
#endif
// Size of buffer to hold descriptors and other data used for enumeration
#define CFG_TUH_ENUMERATION_BUFSIZE 256
#define CFG_TUH_HUB                 1
```
3) We can skip the descriptor files, since as said above a host will receive them and hence must not provide them.
4) In your <chipname>_it.c file (in my case stm32f4xx_it.c), we need to swap the device interrupt handler (tud_int_handler) with that of the host. So we litterally have to change one letter and call 'tuh_int_handler(BOARD_DEVICE_RHPORT_NUM)'. BOARD_DEVICE_RHPORT_NUM is 0 in our case. Return on the next line.
5) tinyUSB expects us to implement two platform specific timing hooks: tusb_time_millis_api (used to be 'board_millis') and tusb_time_delay_ms_api (used to be 'board_delay'). We can add the implementation for these into any .c file that is included in the build. But creating a separate board.c file is probably the cleanest. 
```
--- board.c ---
#include "stm32f4xx_hal.h"
#include "tusb.h"
uint32_t tusb_time_millis_api(void)
{
    return HAL_GetTick();
}
void tusb_time_delay_ms_api(uint32_t ms)
{
    HAL_Delay(ms);
}
```
6) Build and flash. If you have trouble double check or simply try to flip your D+ and D- cables. Since your USB device may be recognized but won't enumerate when they are flipped (happened to me).

**Reply by TobiasHofer** [reply]
to whom it may concern, 
i have just tried to use this guide to implement tinyusb on an stm32g4 and had to add some code to the stm32g4xx_it.c in the user code 1 section:
```
void USBWakeUp_IRQHandler(void) {
  tud_int_handler(BOARD_DEVICE_RHPORT_NUM);
}
```
this is necessary because fore some reason CUBE MX doesn't generate it and as far as i can see there is no way to specify it using CUBE MX

**Comment by viteo** [comment]
successfully added library to CubeIDE without HAL generation (for BluePill f103)

- in configuration tool pin GPIOA 11 and 12 Pins with labels USB_DM and USB_DP for example. do not check USB Device
- init USB peripherials manually:
```
void LL_USB_InitPeriph() // hand-made function
{
    LL_RCC_SetUSBClockSource(LL_RCC_USB_CLKSOURCE_PLL_DIV_1_5);
    LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_USB);
    
    LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_GPIOA);
    
    LL_GPIO_InitTypeDef GPIO_InitStruct = { 0 };
    GPIO_InitStruct.Pin = USB_DM_Pin | USB_DP_Pin;
    GPIO_InitStruct.Mode = LL_GPIO_MODE_ALTERNATE;
    GPIO_InitStruct.Speed = LL_GPIO_SPEED_FREQ_HIGH;
    LL_GPIO_Init(USB_DM_GPIO_Port, &GPIO_InitStruct);
    
    NVIC_SetPriority(USB_HP_CAN1_TX_IRQn, NVIC_EncodePriority(NVIC_GetPriorityGrouping(),0, 0));
    NVIC_SetPriority(USB_LP_CAN1_RX0_IRQn, NVIC_EncodePriority(NVIC_GetPriorityGrouping(),1, 0));
    
    NVIC_EnableIRQ(USB_HP_CAN1_TX_IRQn);
    NVIC_EnableIRQ(USB_LP_CAN1_RX0_IRQn);
}
```

- add interrupt handlers in *_it.c
```
/* USER CODE BEGIN 1 */
void USB_HP_CAN1_TX_IRQHandler(void)
{
    tud_int_handler(BOARD_DEVICE_RHPORT_NUM);
}

void USB_LP_CAN1_RX0_IRQHandler(void)
{
    tud_int_handler(BOARD_DEVICE_RHPORT_NUM);
}
/* USER CODE END 1 */
```
- add TinyUSB lib as mentioned above
- PROFIT

**Comment by sidprice** [comment]
I could use some pointers on getting my STM32L433 custom board working with MSC & CDC tinyUSB. I followed the instructions above and I have the project building. The USB global interrupt is enabled and the addition to call the tinyUSB handler were made. However, the interrupt is never raised.

Any pointers to debugging this would be very much appreciated.

**Comment by sidprice** [comment]
@MimitechIndustries @viteo  I could use some help here. I am a first-time user of tinyUSB and at this time I am stuck :(

This is all the log output I see:

00> USBD init on controller 1
00> sizeof(usbd_device_t) = 51
00> sizeof(tu_fifo_t) = 20
00> sizeof(tu_edpt_stream_t) = 36
00> CDC init
00> MSC init

**Reply by viteo** [reply]
there are thousands of reasons why it may not work.
first of all, did you try ST's USB library. confirm that your custom board is working.

**Reply by sidprice** [reply]
Yes, the board hardware is working, it was tested with as a simple CDC class device. I realize "not working" is vague. What I am looking for is a suggestion or two of where to start debugging/looking.
What is expected after the very short log output shown above?
I have stepped through the initialization code for the USB device and none of the setup functions return an error, all appear to be correct.

**Reply by viteo** [reply]
check USB and GPIO clocks are enabled and set, check check USB_CNTR register for interrupt enable bits set.
also, you may look [here ](https://github.com/hathach/tinyusb/blob/master/hw/bsp/stm32l4/family.c) for some hints, maybe something come up to you.

**Comment by sidprice** [comment]
@viteo I have created a new project based upon the msc_cdc example to try to learn more about tinyUSB. TThe first issue I see in my project is that the "USV" bit is not getting set in PWR->CR2. I have search the code and do not find any reference to this. This suggest I am missing some code somewhere.
The example calls "MX_USB_PCD_Init" and then when the usb device task starts, it calls "tud_init(0)".

**Reply by d4v1dp3** [reply]
For anyone who may have this problem
`SET_BIT(PWR->CR2, PWR_CR2_USV);`
or
`HAL_PWREx_EnableVddUSB();`

**Reply by rob-deutsch** [reply]
I think I've been running into exactly the same problem. Found this post on Google. Thank you for the solution! Somewhat surprising that STM32CubeMX doesn't do this automatically with the USB peripheral enabled, but does with its own CDC middleware enabled.

**Comment by ejaaskel** [comment]
I made an example repository integrating the TinyUSB and CDC dual ports example from TinyUSB to STM32 Nucleo F446RE board. It follows the steps from @MimitechIndustries answer and may be useful for someone starting out with TinyUSB on STM32
https://github.com/ejaaskel/stm32_tinyusb

**Reply by jorgespd** [reply]
@ejaaskel @MimitechIndustries 
Do you by chance have data on the maximum transfer rates you were able to achieve?
I have achieved near 16MByte / s using STM's CDC HS library.  I believe the fastest I can achieve with this tinyusb method is ~6MByte/s

**Reply by ejaaskel** [reply]
No, unfortunately I didn't perform transfer rate testing. That would be an interesting thing to try out though if I have time for it. From a hypothetical point of view the results you've gotten sound sensible in my opinion.

**Reply by docsdom** [reply]
Hi,
I'm also interested in tiny_usb on a STM32WB, succeeded to get it work thanks to the example.
I've even been further and made a composite device with CDC, MSC and audio (ua2c_headset example). Almost everithing is working except for one thing: I can not record anything within Audacity running on the PC.
COM port is ok, MSC is ok, Playing from audacity is ok (data received in the borad), but nothing received within audacity when recording. Actually, the cursor doesn't even advance. But I've checked, data is sent over the usb from the board.
If someone could be of any help ...
Thanks a lot

**Comment by f1ac0** [comment]
Hello, and thank you very much for your instructions that helped me for my first project using tinyUSB, after a few other projects using ST stack.
Just one possible tiny improvement : for the last step, instead of inserting a return inside the _it  file, I recommend unchecking "Call HAL handler" inside the NVIC/"Code Generation" settings for this USB interrupt.

I am sharing this first project here : https://github.com/f1ac0/STM32F103_3xUSBCDC2UART and at the same time I am asking for advice because the 3rd CDC interface is not working as expected even though the code is the same as the first two ones : it does not receive or emit any data on the USB CDC side (the UART3 is working fine when swapped with another one) and the application stops working properly whenever it is accessed. Could it be a limitation of the microcontroller itself even though it should support up to 8 bidir EPs ?

**Reply by hathach** [reply]
What is your debug log ? It usually show tinyusb inside info

**Reply by f1ac0** [reply]
I activated the log with "CFG_TUSB_DEBUG        2" and routed it to stm32 UART1 by defining "__io_putchar".
https://kdrive.infomaniak.com/app/share/429801/db106d12-d5ba-42b5-bbe9-45da0a0bfe2c
I then recorded the log and the actions I did : these are prefixed by %%%% inside the log.

The interesting part is probably this during the configuration of the 3rd CDC:
  Open EP 06 with Size = 64
dcd_pma_alloc 509: ASSERT FAILED
  Open EP 86 with Size = 64
dcd_pma_alloc 509: ASSERT FAILED

**Reply by HiFiPhile** [reply]
You met the STM32F103's hardware limit:
![image](https://github.com/user-attachments/assets/d7a05940-e4bc-4c16-8e32-f014cf4a7132)
Every CDC instance needs 2*IN EP and 1*OUT EP

**Reply by f1ac0** [reply]
I documented this PMA memory that is 512 bytes on the F103, and what buffers tinyUSB need to fit inside it : and yes it it was overflowing, hence the ASSERT inside the log.
To keep the buffers for the CDC endpoint I reduced EP0 buffers, and it is now working fine :)
Thank you hathach for your guidance and for tinyUSB !
And HiFiPhile the actual hardware limit that was hit is the PMA memory size.

Edit : to be more precise, the relevant peripheral for the F103 inside the datasheet is not "USB on-the-go full-speed (OTG_FS)" but "Universal serial bus full-speed device interface (USB)" which states the following : "The size of this dedicated buffer memory must be according to the number of endpoints used and the maximum packet size. This dedicated memory is sized to 512 bytes and up to 16 mono-directional or 8 bidirectional endpoints can be used." and "The number of registers is 8, allowing up to 16 mono-directional/single-buffer or up to 7 double-buffer endpoints(a) in any combination. For example the USB peripheral can be programmed to have 4 double buffer endpoints and 8 single-buffer/mono-directional endpoints." Finally this F103 is suitable for 3xCDC ;)

**Reply by BitLoose** [reply]
It would be a godsend if TinyUSB could be integrated in CubeMX!!
STM's USB drivers are scrappy at best, no composite device support, and they´ve not updated them for a decade.
I've been trying to figure out how to get the volume keys to generate data in my H743 Host keyboard driver, which works with all other keys, but there is no documentation anywhere about how.

**Reply by hathach** [reply]
Unfortunately I don't have time and motivation to test and support cubeide.

**Comment by GalatanuM** [comment]
Will there be a tutorial or something for STM32Duino? I am trying but I fail with everything that I do

**Comment by Valeron01** [comment]
I've created a guide to help get TinyUSB working on the Bluepill (STM32F103C6T6) with CubeIDE
I hope it proves useful - feedback and suggestions are welcome!
https://github.com/Valeron01/BluepillTinyUSB

**Reply by GalatanuM** [reply]
Any hope for stm32duino integration? :D
